### ISSUE TITLE
Dual sign-off for blocklist changes

### ISSUE THREAD

**Issue by jvillalobos** [mozilla/addons-server#13089]
Per requirements from Security, we'll need to implement dual sign-off for blocklist changes. This means that some changes will first go to a "pending" state that then can only be approved by a second user. Any user with access to this admin can reject a block change that is pending.

The changes that require dual sign-off are:
- Adding a new block.
- Deleting a block.
- Changing the version range for a block.

Changing the URL, Reason, or legacy state doesn't require dual sign-off.

Visiting the Edit view of a block in a "pending" state should show what the change is:

- `This is a new block, pending sign-off.`
- `This block will be deleted, pending sign-off.`
- `The version range for this block will be changed to Min version: 0 and Max version: 2.5, pending sign-off.`

And it should have an Approve Change button and Reject Change button. The history should show who requested the change. After approval or rejection, those actions should also be logged in the history.

Blocks with pending changes can't be deleted or have their version range changed again. The pending change needs to be resolved first.

We discussed some kind of notification system for this. I think it would make sense to send an email to a list if something is added to the "pending" list, and when it's approved or rejected. We could just create a Blocklist Admin user group and send the email to those users. Thoughts, @wagnerand?

**Comment by eviljeff** [comment]
> Visiting the Edit view of a block in a "pending" state 

This sounds like for a multiple block action like adding new blocks you'd want the blocks created, and then someone would need to approve them individually.  The idea I had in my head was to require the _bulk create_ action to need sign-off.  So all the guids would all be listed in a single page and then the blocks would be created _after_ the sign-off.

Bulk editing existing blocks would work in a similar way.

(Adding or editing a single block would need some special handling)

**Comment by jvillalobos** [comment]
Yeah, what I've been thinking is that approving a change that is part of a bulk operation would take you to the confirmation step of that operation, including all affected IDs. I'm going to update the mocks with what I'm thinking, and we can discuss if it makes sense.

**Comment by eviljeff** [comment]
If the sign-off page was based around the existing multi-block model/form it would look something like: 
![signoff mockup](https://user-images.githubusercontent.com/768592/70902184-26b37d80-1ff4-11ea-9d30-bdb212a0ebdb.png)
Showing a multi-block that updates a single existing block (because the other guids I submitted were invalid or already blocked for that range)

**Comment by wagnerand** [comment]
The proposal sounds good overall, we might want to change the following:

> Blocks with pending changes can't be deleted or have their version range changed again. The pending change needs to be resolved first.

We should allow further changes and just overwrite the previous pending change for that add-on. A common use-case for that is an admin accidentally selecting the wrong version range.

Also, every user should be able to roll-back their changes.

**Comment by jvillalobos** [comment]
> We should allow further changes and just overwrite the previous pending change for that add-on. A common use-case for that is an admin accidentally selecting the wrong version range.

This flow somewhat breaks dual sign-off. If A submits a change, and B modifies it, neither A nor B should be able to sign-off on it, so that increases the number of people that need to be involved. It's safer for B to reject and either A or B can submit a corrected block that then can be signed off again.

> Also, every user should be able to roll-back their changes.

Agree. That's implicit in the proposed flow (a user can reject their own change), but it's good to call out.

**Comment by jvillalobos** [comment]
I created a [new document with the updated mocks](https://docs.google.com/drawings/d/1xmbpxfRq2ZZvaC9rVOILs0rousgJFNj_AUGWoW9WeUg/edit), so they are easier to compare. These are the changes I made:

- A separate list of blocks pending review, which link to the Review Changes page for each. If there's a way to have a single list for all blocks with a filter, that's also fine with me.
- The block history shows the change being made and the approval (shown in the edit page for a single block).
- When a block is pending changes, the edit page doesn't allow edits or deletions. It shows a message indicating there are changes pending review, with a link to the Review Changes page.
- The Review Changes page has three modes: add, edit, delete. The differences are the headers, the message that appears above the buttons, and in the case of "edit" an indication of the previous values for the min and max version.

**Comment by wagnerand** [comment]
> This flow somewhat breaks dual sign-off. If A submits a change, and B modifies it, neither A nor B should be able to sign-off on it, so that increases the number of people that need to be involved.

Fwiw, kinto allows this. 

> It's safer for B to reject and either A or B can submit a corrected block that then can be signed off again.

This is for sure more time-consuming. We should consider finding a balance between safety and efficiency here. Whether it's safer is debatable: A would file a block, B would reject it and submit a corrected one, then A would approve that. While that's technically B's block then, logically it's A's block, since the original filing and info came from them.

**Comment by jvillalobos** [comment]
Normally both A and B would have access to the original data for the block (in the form of a bug report), and the goal of the process is for one to enter the data and the other to verify it. Giving the second person the ability to edit and push the block means that they're able to submit unchecked changes, which breaks the process. Adding the restriction certainly slows down the process in case of input errors, but I think it's necessary for the dual sign-off requirement to be satisfied.

**Comment by jvillalobos** [comment]
Sorry, the previous comment assumed an incorrect flow on Kinto. The way it works (correct me again if I'm wrong) is that A creates the block, B edits, and then A can sign off on the edited version. This assumes B did the review before the edit, which I think is fair (and would need to be part of the documented process).

So, this is more of a shortcut to avoid having to re-enter the corrected block. I'd like to hear from Andrew how much more complexity this introduces. This is what I'm thinking:

- The Review Changes page allows changing the URL, Reason, and URL and Approve Changes without additional sign-off. Doesn't apply to deletions.
- The Review Changes page allows changing the min and max version. If those values change, the user making the change effectively takes over the change request, and someone else will need to approve it. That someone else can be the user who made the original change request. In this case the Approve Changes button should be disabled and an Update button should be enabled instead. The message at the bottom should read `This will override the current change and someone else will need to sign off on it.` Doesn't apply to deletions.
- The list of IDs can't be changed.

**Comment by eviljeff** [comment]
> A separate list of blocks pending review, which link to the Review Changes page for each. If there's a way to have a single list for all blocks with a filter, that's also fine with me.

Is it okay to have the delete action be a separate list?  I've not planned how delete sign-off would be implemented but it's possible that keeping the sign-off models for delete and add/edit seperate will be more straightforward.

> The Review Changes page has three modes: add, edit, delete. The differences are the headers, the message that appears above the buttons,

It's not quite as clean-cut between adding and editing - when you input a list of guids there could be a mix between guids that don't exist as blocks (add) and ones that do (edit), and they would all be shown as a single action because they've been entered together (at least, that's how I interpreted it from the previous issues about multi-add-on blocks).  So the add and edit views would be the same view.

>  and in the case of "edit" an indication of the previous values for the min and max version. 

The blocks that already exist could have a range of different previous values, so showing a single "Previous Value" next to the new values isn't possible.  Though I planned to include some of the details next to each guid (like the mockup I posted above).

re: delete mock - the page shows min version, max version, url and reason fields but they won't exist because that's not entered for a delete action, plus like add/edit, the blocks being deleted could have a range of different values.  Though if you want some of the details could be included next to each guid like add/edit.

**Comment by eviljeff** [comment]
> I'd like to hear from Andrew how much more complexity this introduces.

In general, the more places something can be edited, the more edge cases are introduced and the more (potentially) unstable the resulting system will be.  Simple edits that don't lead to further sign-offs don't sound too problematic (just need to make sure the logged is accurate regarding who entered what information).  

> The Review Changes page allows changing the min and max version

So, the biggest problem I can see here is that the min and max version _determines_ which existing blocks are updated and which aren't in the input step.  If you change the min/max version then that evaluation maybe isn't correct any longer (i.e. a guid user A entered and was skipped over as already blocked now _should_ be updated, and visa-versa).  If the sign-off view then has to handle that too then we've created a combined add&edit&signoff view (triple the complexity!).  

Really, it'd be a lot cleaner if min/max version changes were prohibited in the sign-off view and required a rejection and a new add/edit action.

**Comment by jvillalobos** [comment]
> Is it okay to have the delete action be a separate list? 

Yes

> when you input a list of guids there could be a mix between guids that don't exist as blocks (add) and ones that do (edit), and they would all be shown as a single action

Hadn't thought about that, but I see what you mean.

> The blocks that already exist could have a range of different previous values, so showing a single "Previous Value" next to the new values isn't possible. Though I planned to include some of the details next to each guid (like the mockup I posted above).

Okay.

> re: delete mock - the page shows min version, max version, url and reason fields but they won't exist because that's not entered for a delete action, plus like add/edit, the blocks being deleted could have a range of different values. Though if you want some of the details could be included next to each guid like add/edit.

> Really, it'd be a lot cleaner if min/max version changes were prohibited in the sign-off view and required a rejection and a new add/edit action.

Agree. I'm okay with having that restriction, to keep thing easier to maintain. It means a bit more work when amending blocks, but that shouldn't be so common.

Yes, including the detail would be good.

**Comment by jvillalobos** [comment]
One more thing we discussed with the Security people: there could be a threshold for requiring dual sign-off. If all add-ons involved in a block have user counts under that threshold, the block change would be allowed without sign-off. We need to figure out what the right number is, but we talked about 100K being a good starting point.

Since we don't have stats for unlisted extensions at the moment, any ID without stats should required dual sign-off. Once we have complete data this threshold should become more practical.

**Comment by wagnerand** [comment]
Another requirement that just came up yesterday is that the review workflow should have a separate permission, as we might have users who are only reviewing blocks.

**Comment by eviljeff** [comment]
It's already quite complicated code so the basic dual sign-off will be implemented first, then anything extra in a separate PR/issue.

**Comment by AlexandraMoga** [comment]
Do we have a final requirements document for this feature? It would be nice to see the information consolidated somewhere since it's becoming quite difficult to understand the expected behavior of the product just by reading the exchanged comments.

**Comment by eviljeff** [comment]
The major dual-sign off functionality is complete now so feel free to give it a try on stage/dev 😸 

There are some outstanding bugs/feature requests in https://github.com/orgs/mozilla/projects/112 that will be worked on over the remainder of Q1 but I don't believe any of them are blockers for the MVP (I marked anything non-MVP as P4; and P3 for MVP - the stats P3 is blocked by the stats project though).

Please file new issues for any missing/incorrect functionality and add them to the project.

QA: the individual issues have been verified as they've landed so hopefully there isn't anything extra but it would be good to verify the complete flow now it's complete-ish.
